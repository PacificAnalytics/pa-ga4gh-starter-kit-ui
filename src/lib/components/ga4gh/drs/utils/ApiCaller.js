import axios from 'axios';

const ApiCaller = (requestConfig, handleResponseData, handleError) => {
  axios(requestConfig)
  .then(response => {
    handleResponseData(response.data);
  })
  .catch(error => {
    console.log('caught the error');
    console.log(error);
    console.log(handleError);
    console.log("***");
    handleError(error)
  })
}

export default ApiCaller;
