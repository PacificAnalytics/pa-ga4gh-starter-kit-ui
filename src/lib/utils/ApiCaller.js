import axios from 'axios';

const ApiCaller = (requestConfig, handleResponseData, handleError) => {
  axios(requestConfig)
  .then(response => {
    handleResponseData(response.data);
  })
  .catch(error => {
    handleError(error)
  })
}

export default ApiCaller;
