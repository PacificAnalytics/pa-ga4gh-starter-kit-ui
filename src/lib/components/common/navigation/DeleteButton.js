import React, { useState } from 'react';
import {
    Typography,
    Button, 
    Dialog, 
    DialogTitle, 
    DialogContent, 
    DialogContentText, 
    DialogActions
} from '@material-ui/core';
import { Delete } from '@material-ui/icons';
import { useHistory } from 'react-router-dom';
import ApiCaller from '../../../utils/ApiCaller';
import { scrollToTop } from '../../../functions/common';
import deleteButtonStyles from '../../../styles/common/navigation/deleteButtonStyles';

const DeleteButton = props => {
    let history = useHistory();
    const [confirmationDialogIsOpen, setConfirmationDialogIsOpen] = useState(false);
    const closeConfirmationDialog = () => setConfirmationDialogIsOpen(false);
    const deleteEntity = () => {
        let requestConfig = {url: props.deleteURL, method: 'DELETE'}
        ApiCaller(
            requestConfig,
            responseData => {
                props.setSuccessMessageFunc(`${props.entityName} '${props.id}' has been successfully deleted`);
                props.updateDataFunc()
                history.push(props.redirect);
                scrollToTop();
            },
            error => {
                props.setErrorFunc(error);
            }
        )
        setConfirmationDialogIsOpen(false);
    }

    const classes = deleteButtonStyles();
    return (
        <div className={classes.div}>
            <Button
                className={classes.root}
                variant="outlined"
                onClick={() => setConfirmationDialogIsOpen(true)}
            >
                Delete
                <Delete />
            </Button>
            <Dialog open={confirmationDialogIsOpen} onClose={closeConfirmationDialog}>
                <DialogTitle align='center'>
                    {`Are you sure you want to delete this ${props.entityName}`}?
                </DialogTitle>
                <DialogContent>
                    <DialogContentText align='center'>
                        {`Deleting this ${props.entityName} will permanently remove it from the database.`}
                    </DialogContentText>
                </DialogContent>
                <DialogActions>
                    <Button variant='contained' color='primary' onClick={closeConfirmationDialog}>
                        <Typography variant='button'>Cancel</Typography>
                    </Button>
                    <Button variant='contained' color='primary' 
                    onClick={e => deleteEntity() }>
                        <Typography variant='button'>Delete</Typography>
                    </Button>
                </DialogActions>
            </Dialog>
        </div>
    )
}

export default DeleteButton;
