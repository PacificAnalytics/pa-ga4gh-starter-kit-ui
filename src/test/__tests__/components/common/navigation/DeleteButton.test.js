import React from 'react';
import renderer from 'react-test-renderer';
import {getByRole, getByText, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import {mockValidTestBlob} from '../../../../resources/MockData';
import DeleteButton from '../../../../../lib/components/common/navigation/DeleteButton';
import {
    setSuccessMessage, 
    retrieveDrsObjectsList, 
    setError
} from '../../../../resources/MockFunctions';
import ApiCaller from '../../../../../lib/utils/ApiCaller';

jest.mock('../../../../../lib/utils/ApiCaller');


let deleteMockFunctions = null;

beforeEach(() => {
    deleteMockFunctions = {
        setSuccessMessage: setSuccessMessage,
        retriveDrsObejctsList: retrieveDrsObjectsList,
        setError: setError,
    }
});

afterEach(() => {
    setSuccessMessage.mockClear();
    retrieveDrsObjectsList.mockClear();
    setError.mockClear();
})

test('<DeleteButton /> can delete DRS Object successfully', () => {
    ApiCaller.mockImplementation(() => {
        setSuccessMessage(`DRS Object '${mockValidTestBlob.id}' has been successfully deleted`);
        retrieveDrsObjectsList();
    });
    
    let {container} = render(
        <DeleteButton
            entityName='DRS Object'
            id={mockValidTestBlob.id}
            {...deleteMockFunctions}
        />
    );
    expect(container.firstChild).toMatchSnapshot();
    let deleteButton = screen.getByRole('button');
    expect(deleteButton).toBeInTheDocument();
    userEvent.click(deleteButton);
    let confirmationDialog = screen.getByRole('dialog');
    expect(confirmationDialog).toBeVisible();
    expect(getByText(confirmationDialog, 'Are you sure you want to delete this DRS Object?')).toBeInTheDocument();
    expect(getByText(confirmationDialog, 'Deleting this DRS Object will permanently remove it from the database.')).toBeInTheDocument();
    expect(getByRole(confirmationDialog, 'button', {name: 'Cancel'})).toBeInTheDocument();
    expect(getByRole(confirmationDialog, 'button', {name: 'Delete'})).toBeInTheDocument();

    // clicking cancel button should close dialog
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Cancel'}));
    expect(confirmationDialog).not.toBeVisible();

    // reopening dialog and clicking delete button should delete DRS Object
    userEvent.click(deleteButton);
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Delete'}));
    expect(setSuccessMessage).toHaveBeenCalled();
    expect(retrieveDrsObjectsList).toHaveBeenCalled();
    expect(setError).not.toHaveBeenCalled();
});

test('<DeleteButton /> can handle errors', () => {
    ApiCaller.mockImplementation(() => {
        setError('Error: DRS object was not deleted successfully')
    });
    
    let {container} = render(<DeleteButton id={mockValidTestBlob.id} {...deleteMockFunctions}/> );
    let deleteButton = screen.getByRole('button');
    userEvent.click(deleteButton);
    let confirmationDialog = screen.getByRole('dialog');
    expect(confirmationDialog).toBeVisible();
    // clicking delete button should cause error
    userEvent.click(deleteButton);
    userEvent.click(getByRole(confirmationDialog, 'button', {name: 'Delete'}));
    expect(setSuccessMessage).not.toHaveBeenCalled();
    expect(retrieveDrsObjectsList).not.toHaveBeenCalled();
    expect(setError).toHaveBeenCalled();
});
