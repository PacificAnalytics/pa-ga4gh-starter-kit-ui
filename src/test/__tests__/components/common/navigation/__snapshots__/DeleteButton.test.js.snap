// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteButton /> can delete DRS Object successfully 1`] = `
<div
  class="makeStyles-div-1"
>
  <button
    aria-label="delete-button"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-root-2"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Delete
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;
